src/sounds.cpp

```diff
@@ -388,16 +393,790 @@ std::pair<std::vector<tripoint>, std::vector<tripoint>> sounds::get_monster_soun
+        play_ambient_variant_sound( "environment", "indoors_rain", get_heard_volume( g->u.pos() ), 4,
+                                    1000 );
+    }
+    if( ( !g->is_sheltered( g->u.pos() ) && g->weather != WEATHER_CLEAR
+            && is_channel_playing( 5 ) != 1  &&
+            is_channel_playing( 6 ) != 1  && is_channel_playing( 7 ) != 1  && is_channel_playing( 8 ) != 1
+            &&
+            is_channel_playing( 9 ) != 1  && !g->u.get_effect_int( "deaf" ) > 0 )
+            || ( !g->is_sheltered( g->u.pos() ) &&
+                 g->weather != previous_weather  && !g->u.get_effect_int( "deaf" ) > 0 ) ) {
+        fade_audio_group( 1, 1000 );
+        // We are outside and there is precipitation
+        switch( g->weather ) {
+	case WEATHER_ACID_DRIZZLE:        
+	case WEATHER_DRIZZLE:
+            play_ambient_variant_sound( "environment", "WEATHER_DRIZZLE", get_heard_volume( g->u.pos() ), 9,
+                                        1000 );
+            break;
+        case WEATHER_RAINY:
+            play_ambient_variant_sound( "environment", "WEATHER_RAINY", get_heard_volume( g->u.pos() ), 8,
+                                        1000 );
+            break;
+	case WEATHER_ACID_RAIN:
+	case WEATHER_THUNDER:
+        case WEATHER_LIGHTNING:
+            play_ambient_variant_sound( "environment", "WEATHER_THUNDER", get_heard_volume( g->u.pos() ), 7,
+                                        1000 );
+            break;
+        case WEATHER_FLURRIES:
+            play_ambient_variant_sound( "environment", "WEATHER_FLURRIES", get_heard_volume( g->u.pos() ), 6,
+                                        1000 );
+            break;
+	// I didn't know where to put these or even if the first two were needed. They probably shouldn't make any noise, but I was getting an error when compiling when I left them out.
+	case WEATHER_NULL:
+	case NUM_WEATHER_TYPES:        
+	case WEATHER_CLEAR:
+	case WEATHER_SUNNY:
+	case WEATHER_CLOUDY:
+        case WEATHER_SNOWSTORM:
+        case WEATHER_SNOW:
+            play_ambient_variant_sound( "environment", "WEATHER_SNOW", get_heard_volume( g->u.pos() ), 5,
+                                        1000 );
+            break;
+        }
+    }
+    // Keep track of weather to compare for next iteration
+    previous_weather = g->weather;
+}
+
+int sfx::get_heard_volume( const tripoint source ) {
+    int distance = rl_dist( g->u.pos3(), source );
+    // fract = -100 / 24
+    const float fract = -4.166666;
+    int heard_volume = fract * distance - 1 + 100;
+    if( heard_volume <= 0 ) {
+        heard_volume = 0;
+    }
+    heard_volume *= g_sfx_volume_multiplier;
+    return ( heard_volume );
+}
+
+int sfx::get_heard_angle( const tripoint source ) {
+    int angle = g->m.coord_to_angle( g->u.posx(), g->u.posy(), source.x, source.y ) + 90;
+    //add_msg(m_warning, "angle: %i", angle);
+    return ( angle );
+}
+
+void sfx::generate_gun_soundfx( const tripoint source ) {
```

Comment:
If you move the call to `play_variant_sound` from `Creature::projectile_attack` into `player::fire_gun`, you could forward the firing creature to the function, so it does not have to get this information from the map again. You can also forward all the relevant data, like what item is used to fire and so on.

This allows to get rid of the special cases for gun items used by the various monattack functions. The code can easily get out of sync, someone adds a new monattack function using a new gun, or changes the a monattack function to use a different gun and may not update the code here.

The code is very fragile anyway. The gun item used by a monster attack is defined in the monattack functions, but each and every monster can have each and every monattack function. Just because the monster is of type "mon_turret_bmg" doesn't mean it will fire a "m2browning", anyone could change the JSON data to let it fire a laser riffle instead.